### PR TITLE
[#21] [#22] [Integration] As a user, I can load more or refresh surveys on the home screen

### LINKED ISSUE #21: [Integration] As a user, I can load more surveys on the home screen
## Why

The user can see a list of the surveys which have not been finished by the users on the home screen. The user can load more items when swiping to the end of the list.

## Acceptance criteria

When the user swipes to the 7th item of the current page, call the API `/api/v1/surveys` to get surveys of the next page.

## Resources

- [API document for /api/v1/surveys](https://documenter.getpostman.com/view/11835486/T1LJjTqj#2dd12ea2-3ecb-46d5-a48b-64256346e4e1)

### LINKED ISSUE #22: [Integration] As a user, I can pull right to refresh the survey list on the home screen
## Why

The user can see a list of the surveys which have not been finished by the users on the home screen. The user can pull right to reload the list to see any updates.

## Acceptance criteria

On the home screen, if the user pulls right when it shows the first item of the survey list, refresh the survey list.

Show the loading indicator while refreshing.

## Resources

- [API document for /api/v1/surveys](https://documenter.getpostman.com/view/11835486/T1LJjTqj#2dd12ea2-3ecb-46d5-a48b-64256346e4e1)

### PR DESCRIPTION
**Pull request by nmint8m**
- Close #21 #22

## What happened 👀

The user can see a list of the surveys which have not been finished by the users on the home screen. 

The user can load more items when swiping to the end of the list.

The user can pull down to reload the list to see any updates.

## Insight 📝

I have implemented 2 styles of refresh:

- Pull down to refresh
- Swipe right to refresh

In the POW, I pulled down to refresh.

### CODE REVIEW

File: lib/views/home/home_view.dart [hunk 1]
```diff
@@ -36,8 +38,8 @@ class HomeViewState extends ConsumerState<HomeView> {
         height: MediaQuery.of(context).size.height,
         child: Consumer(builder: (_, ref, __) {
           final index = ref.watch(focusedItemIndexStream).value ?? 0;
-          final surveyList = ref.watch(surveyListStream).value ?? [];
-          return surveyList.isEmpty
+          final surveyList = ref.read(surveyListStream).value ?? [];
```
Comment (manh-t) [inline lib/views/home/home_view.dart]: @nmint8m it recommend to use `watch` in Consumer, avoid `read` amap 💩

File: lib/views/home/home_view.dart [hunk 2]
```diff
@@ -55,39 +57,47 @@ class HomeViewState extends ConsumerState<HomeView> {
...
+            count: length,
+          ),
+          const SizedBox(height: 160),
         ],
       );
 
+  Widget get _mainBody => Consumer(
+        builder: (_, ref, __) {
+          final surveyList = ref.watch(surveyListStream).value ?? [];
+          final index = ref.read(focusedItemIndexStream).value ?? 0;
```
Comment (manh-t) [inline lib/views/home/home_view.dart]: same here, please change to `watch`

File: lib/views/home/survey_section/survey_list.dart [hunk 3]
```diff
@@ -2,26 +2,100 @@ import 'package:flutter/material.dart';
...
-    return PageView.builder(
-      scrollDirection: Axis.horizontal,
-      controller: itemController,
-      onPageChanged: (index) => onItemChange.value = index,
-      itemCount: surveyList.length,
-      itemBuilder: (context, index) => SurveyCell(surveyList[index]),
+    _isLoading = false;
+    switch (refreshStyle) {
+      case RefreshStyle.swipeRightToRefresh:
+        return _buildSwipeRightToRefreshPageView(context);
```
Comment (manh-t) [inline lib/views/home/survey_section/survey_list.dart:38]: 🙏 🙏 🙏 🙏

File: lib/views/home/survey_section/survey_list.dart [hunk 4]
```diff
@@ -2,26 +2,100 @@ import 'package:flutter/material.dart';
...
+        itemBuilder: (context, index) => SurveyCell(surveyList[index]),
+      ),
+    );
+  }
+
+  Widget _buildPullDownToRefreshPageView(BuildContext context) {
+    final screenWidth = MediaQuery.of(context).size.width;
+    itemController.addListener(() {
+      if (itemController.offset > screenWidth * (surveyList.length - 1) + 10 &&
+          !_isLoading) {
```
Comment (manh-t) [inline lib/views/home/survey_section/survey_list.dart]: I think we can load more items if we swipe to near the end of the list (maybe 2-3 items remaining), it's more simple IMO 😄

Comment (nmint8m) [reply lib/views/home/survey_section/survey_list.dart]: Changed to `- 3` 👍

File: lib/views/home/home_view.dart (hunk 2 above)
Comment (nmint8m) [reply lib/views/home/home_view.dart]: @manh-t 

I want to rebuild the SurveyList whenever `surveysStream` gets a new value. If I change to `watch`, which means the SurveyList will be rebuilt when `surveysStream` or `focusedItemIndexStream` get new values, but it doesn't work that way.

I will move it into the Future callback block, as you mentioned in the chat thread.

**Review by manh-t — APPROVED** [review]

**Pull request merged**